### DISCUSSION TITLE
GraphQL / Postgraphile?

### DISCUSSION THREAD

**Discussion by rattrayalex** [supabase/supabase#1057] [Feature Requests] [answered]
Hi there,

I don't use supabase, but am considering it for future projects. 
I was a bit surprised in the choice of PostgREST over Postgraphile as the API layer, as the latter generates a GraphQL api which is more popular for building frontends these days from what I can tell. But of course no reason not to include both. 

Are you considering adding a GraphQL layer in the future? If so, would it likely be [Postgraphile](https://www.graphile.org/postgraphile/), or something else?

If not, curious why not? (I don't have a strong opinion myself, just curious)

Cheers,
Alex

**Comment by kiwicopple** [comment]
> Are you considering adding a GraphQL layer in the future?

Not at this stage. I know it seems quick and easy to add a tool, however every time we add a tool to the stack (or release a new feature) we create a considerable amount of work for the team - security, docs, UI, benchmarking, support, libraries, community management around the libraries, employing maintainers, sponsoring etc. 

We still have a lot of core functionality to add (Workflows, Functions, offline-support on our libraries etc). Perhaps after that we can revisit, and then reach out to the GraphQL communities to see if they are interested in collaborating

**Reply by rattrayalex** [reply]
Makes sense, thanks!

**Comment by ozum** [comment]
Second this. @kiwicopple, you are right it isn't easy to add a tool especially a monstrous feature like GraphQL considering:

- Authentication integration,
- Security (authorization, rate limiter, calculating how heavy a query is etc., Kong could not help here for GraphQL I guess)
- Supabase storage integration

and some more.

However, I like to see an open-source integration like Postgraphile integration in the roadmap even in the distant future.

**Reply by rattrayalex** [reply]
Agreed – fwiw, I would not advise integrating PostGraphile until their upcoming V5 has been released, as it will include major changes to the API.

**Comment by karolzlot** [comment]
I would like to see something written in Rust for GraphQL, I think PostGraphile may not be fast enough for some use cases.

**Reply by rattrayalex** [reply]
What kind of use cases? Have you benchmarked those cases?

Postgraphile's architecture unlocks some very impressive performance; I don't think a faster language runtime would make much of a difference.

**Reply by karolzlot** [reply]
@rattrayalex 

I didn't do any benchmark. I hope you are right about this.

----

Hasura sometimes isn't fast enough as I remember I read in one GitHub issue in past.

**Reply by olirice** [reply]
We're investigating speeding up the hotspots in our graphql extension, [pg_graphql](https://github.com/supabase/pg_graphql), in rust

+1 wrt PostGraphile being very performant

**Reply by Sleepful** [reply]
Postgraphile is much faster than Hasura or Prisma, it doesn't suffer N+1 issue to begin with.

**Comment by philohelp** [comment]
What's interesting about adding a graphQL layer to supabase is that the choice of the database for the app becomes secondary. So you wouldn't have to worry so much about extending support to other databases.

**Reply by karolzlot** [reply]
Could you explain why do you think so?

**Reply by philohelp** [reply]
I chose supabase for my new project although I come from a noSQL + graphQL background. The decision would have been easier with a graphQL API directly available. It's a bit less work to do.

**Comment by jaredramirez** [comment]
I was able to get this working straightforwardly with following steps:

1. Setup a basic JS server
2. Pull the JWT off the `Authorization` header and verify it with the same secret Supabase uses (You can get you secret from the Supabase dash). This is typically done in a middleware with most JS server frameworks
3. In the middleware after verifying, store the JWT claims on the request somewhere (ex: `req.jwtClaims = ...`)
4. Setup your postgraphile config like:
```javascript
let postgraphileConfig = {
  // put the rest of your config here
  // https://www.graphile.org/postgraphile/usage-library/#recommended-options

  pgSettings: (req) => {
    let jwtClaims = req.jwtClaims
    jwtClaims = jwtClaims.sub && jwtClaims.email && jwtClaims.role ? jwtClaims : null;

    return {
      role: jwtClaims ? jwtClaims.role : "anon",
      ["request.jwt.claim.sub"]: jwtClaims ? jwtClaims.sub : undefined,
      ["request.jwt.claim.email"]: jwtClaims ? jwtClaims.email : undefined,
      ["request.jwt.claim.role"]: jwtClaims ? jwtClaims.role : undefined,
      // TODO Are there other things we need to set for supabase?
      // TODO: Update this for Postgres 14: https://github.com/supabase/supabase/discussions/4169
    };
  },
};
```
5. Setup postgraphile like:
```javascript
const Koa = require('koa');
const { postgraphile: postgraphileMiddleware } = require("postgraphile");

let postgraphileConfig = {
  // from step 4
};

let app = new Koa();
app.use(customAuthMiddleware) // from step 2
app.use(postgraphileMiddleware(process.env.CONN, ["public"], config))

app.listen(process.env.PORT);
```

**Reply by sseppola** [reply]
Did you hook into the existing supabase roles? specifically `authenticator` in order to assume roles the way postgrest does.
I cannot find the authenticator password in order to do this.

**Reply by jaredramirez** [reply]
In the postgraphile config, this line `role: jwtClaims ? jwtClaims.role : "anon",` sets the role. The role decoded in the JWT should typically be `authenticated`. As far as using the `authenticator` role, I didn't use it.

**Reply by sseppola** [reply]
I mean, how do you enable postgraphile to grant these roles in incoming requests? How does postgraphile authenticate with postgres?

**Reply by jaredramirez** [reply]
I setup postgraphile to use admin credentials, then I decode/verify the JWT in middleware. After in `pgSettings`, after pulling the `role` from the decoded JWT, postgraphile calls `set role ${role}` so it doesn't use admin credentials when processing user requests.

**Reply by sseppola** [reply]
Ah I see, thanks!

**Comment by olirice** [comment, marked as answer]
Supabase now has builtin support for GraphQL https://supabase.com/blog/2022/03/29/graphql-now-available

**Comment by adimyth** [comment]
I had built a Supabase + GraphQL stack for one of my projects, it was super convenient to use & reduced the need for a backend completely. You can read about it here - [Migrate your MySQL database to Postgres with Supabase, Hasura & pgloader](https://medium.com/@adi_myth/migrate-your-mysql-database-to-postgres-with-supabase-hasura-pgloader-3ef63bedd38)

**Reply by jtkeyva** [reply]
@adimyth thanks. How is the graphql realtime performace working for you? If you are using multiple tables in a query, is it still realtime? Curious how you feel about it now after some time using it. You still using Hasura for this?
